MekHQ/src/mekhq/service/MassRepairService.java

```diff
@@ -179,85 +210,218 @@ public int compare(Unit o1, Unit o2) {
 	}
 
-	public static int performUnitMassRepairOrSalvage(CampaignGUI campaignGUI, Unit unit, boolean isSalvage,
-			List<MassRepairOption> mroList, MassRepairConfiguredOptions configuredOptions) {
-		String actionDescriptor = isSalvage ? "salvage" : "repair";
-		Campaign campaign = campaignGUI.getCampaign();
+	private static String generateUnitRepairSummary(String baseDescription,
+			Map<Integer, List<MassRepairUnitAction>> unitActionsByStatus, int status) {
 
-		campaign.addReport(String.format("Beginning mass %s on %s.", actionDescriptor, unit.getName()));
+		if (!unitActionsByStatus.containsKey(status)) {
+			return "";
+		}
 
-		ArrayList<Person> techs = campaign.getTechs(true);
+		int count = unitActionsByStatus.get(status).size();
+
+		return String.format(baseDescription, count, count == 1 ? "" : "s");
+	}
+
+	private static void generateCampaignLogForUnitStatus(Map<Integer, List<MassRepairUnitAction>> unitActionsByStatus,
+			int status, String statusDesc, CampaignGUI campaignGUI) {
+		if (!unitActionsByStatus.containsKey(status) || unitActionsByStatus.get(status).isEmpty()) {
+			return;
+		}
+
+		StringBuilder sbMsg = new StringBuilder();
+		sbMsg.append(statusDesc);
+
+		List<MassRepairUnitAction> unitsByStatus = unitActionsByStatus.get(status);
+
+		for (MassRepairUnitAction mrua : unitsByStatus) {
+			sbMsg.append("<br/>- " + mrua.getUnit().getName());
+		}
+
+		campaignGUI.getCampaign().addReport(sbMsg.toString());
+	}
+
+	public static MassRepairUnitAction performUnitMassRepairOrSalvage(CampaignGUI campaignGUI, Unit unit,
+			boolean isSalvage, List<MassRepairOption> mroList, MassRepairConfiguredOptions configuredOptions) {
+		Campaign campaign = campaignGUI.getCampaign();
 
-		int totalActionsPerformed = 0;
+		ArrayList<Person> techs = campaign.getTechs(true);
 
 		if (techs.isEmpty()) {
-			campaign.addReport(String.format("<font color='red'>No available techs to %s parts %s %s.</font>",
-					actionDescriptor, isSalvage ? "from" : "on", unit.getName()));
-		} else {
-			// Filter our tech list to only our techs that can work on this unit
-			for (int i = techs.size() - 1; i >= 0; i--) {
-				Person tech = techs.get(i);
+			return new MassRepairUnitAction(unit, isSalvage, MassRepairUnitAction.STATUS.NO_TECHS);
+		}
 
-				if (!tech.canTech(unit.getEntity())) {
-					techs.remove(i);
-				}
-			}
+		MassRepairUnitAction unitAction = new MassRepairUnitAction(unit, isSalvage,
+				MassRepairUnitAction.STATUS.NO_ACTIONS);
 
-			Map<Integer, MassRepairOption> mroByTypeMap = new HashMap<Integer, MassRepairOption>();
+		// Filter our tech list to only our techs that can work on this unit
+		for (int i = techs.size() - 1; i >= 0; i--) {
+			Person tech = techs.get(i);
 
-			for (int i = 0; i < mroList.size(); i++) {
-				mroByTypeMap.put(mroList.get(i).getType(), mroList.get(i));
+			if (!tech.canTech(unit.getEntity())) {
+				techs.remove(i);
 			}
+		}
```

Comment:
It's generally a bad idea to modify a `Collection` while you're iterating over it.  Better options would be either
1. Create an `availableTechs` `Collection` to contain the techs you want to use or
2. Create a `removeTechs` `Collection` and then call `techs.removeAll(removeTechs)` after the loop.

Reply:
I disagree. For an ArrayList (which this is) arrayCopy is very very fast. It's a very basic level function. If you iterate from the back of the list and remove those that you don't need, it'll be extremely fast and you don't end up with index collisions. Additionally, creating a list of techs to remove is minimally impactful but possibly unnecessary if we don't even need to remove any of the techs. We're creating a list for no reason and doing additionally checks on it for the removeAll().